### PR TITLE
use capacitor for building native apps with webview

### PR DESCRIPTION
**Pull request by norbertschuler**
To be able to put our web application into the Apple and Google PlayStore we need to encapsulate them into native apps. https://capacitorjs.com/ is one possible solution, but as it puts all the code locally inside the apps, it would demand regularly do releases of the native apps whenever we change the web application.

To support Auth0 the apps open a browser view and need to listen to the callback URLs to open them again in the native apps by using a custom scheme `org.pftp.app://` which a custom Auth0 application.

Build the app by running `npm run build-mobile` and then build a iOS with Xcode or Android app with Android Studio. You have to use a custom `.env` for this (ask me for a sample config) as the following env variables differ:

```
AUTH0_CUSTOM_DOMAIN=accounts.plant-for-the-planet.org
AUTH0_CLIENT_ID=6MME4DhfTZes2myvI0NXgWqGSIZdy0IH
NEXTAUTH_URL=org.pftp.app://www1.plant-for-the-planet.org
```

Changes in this pull request:
- added capacitor for building iOS and Android native apps

TODO:
- [x] add automatic build process, currently using a custom .env.production.local file to build the iOS/Android app

Depends on #1558 to display the the app correctly on mobile devices with a notch.

### CODE REVIEW

File: src/features/common/Layout/UserPropsContext.tsx [hunk 1]
```diff
@@ -25,15 +26,39 @@ function UserPropsProvider({ children }: any): ReactElement {
...
+  const [callbackLoaded, setCallbackLoaded] = React.useState(false);
+
+  React.useEffect(() => {
+    App.addListener('appUrlOpen', async (event: URLOpenListenerEvent) => {
+      // just a test: event.url = event.url.replace('org.pftp.app', 'https')
+      console.log("event.url", event.url)
+      // Example url: https://www1.plant-for-the-planet.org/login?code=***&state=***
+      if (event.url.includes('state') && (event.url.includes('code') || event.url.includes('error'))) {
+        await handleRedirectCallback(event.url).then(() => {
+          setCallbackLoaded(true)
```
Comment (norbertschuler) [inline src/features/common/Layout/UserPropsContext.tsx:48]: This does not work in iOS.

**Comment by sagararyal** [comment]
#### 📝 Changed routes:


| [**/claim/\[type\]/\[code\]**](https://planet-webapp-27bqpq1do-planetapp.vercel.app/claim/[type]/[code]) | [**/login**](https://planet-webapp-27bqpq1do-planetapp.vercel.app/login) |
|:-:|:-:|
|<a href="https://planet-webapp-27bqpq1do-planetapp.vercel.app/claim/[type]/[code]">Screenshot of /claim/[type]/[code]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/AC1pWOt0MsaziwOWzT2ve_a0Ag2UwXh9ydyW3yKxv3AHzkX1GrU2dk2TLAgfrhsXusS1cW3LsJpXtzwJerhuDg.png">(view full size)</a> |<a href="https://planet-webapp-27bqpq1do-planetapp.vercel.app/login">Screenshot of /login</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/MlZ5Ws5GRLf7diNnGx0rVhqVXx_i5Jez0qNY48iVfpzQIZyBHVv2EDZvyvYbzsvBaox-irbLnfCiiSMpiqpVOA.png">(view full size)</a> |



| [**…e/projects/new-project**](https://planet-webapp-27bqpq1do-planetapp.vercel.app/profile/projects/new-project) |
|:-:|
|<a href="https://planet-webapp-27bqpq1do-planetapp.vercel.app/profile/projects/new-project">Screenshot of /profile/projects/new-project</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/feyViU1nchbWDomtOxiXJIzkuGYzyavnv0g2Mg0flv6dYO2KerOkZ0pRIhPn9uf2yOAzk0bnPFZgVFPbvb7B7w.png">(view full size)</a> |







Commit b17f9f90d0014e91e7e4fd54ae66f3b4b426e95f (https://planet-webapp-27bqpq1do-planetapp.vercel.app).

File: src/features/common/Layout/UserPropsContext.tsx [hunk 2]
```diff
@@ -13,8 +15,8 @@ export const UserPropsContext = React.createContext({
   token: null,
   isLoading: true,
   isAuthenticated: false,
-  loginWithRedirect: ({}) => {},
-  logoutUser: (value: string | undefined) => {},
+  loginWithRedirect: (redirectUri: string | undefined, ui_locales: string | undefined) => {},
+  logoutUser: (returnUrl: string | undefined) => {},
```
Comment (mohitb35) [inline src/features/common/Layout/UserPropsContext.tsx:19]: I wonder if there's a way to remove the TS warning we receive here as the variables are not initially used in the empty function.

**Comment by mohitb35** [comment]
This seems to break the signup process. After signing up as a new user with email and password, I got stuck on 

[](https://planet-webapp-rfqi8ym5d-planetapp.vercel.app/login?error=unauthorized&error_description=401&state=QXJfeVhkaC15R3g0SHJRVHVrckp5dkRNM0VrTlZzMmJYMmhDOWFLLTBGNQ%3D%3D)

instead of being directed to [/verify-email](url).

I suspect that this code in `src/features/common/Layout/Navbar/index.tsx` is not running as expected.  

```
if (auth0Error) {
    if (auth0Error.message === '401') {
      if (typeof window !== 'undefined') {
        setUser(null);
        logoutUser(`${process.env.NEXTAUTH_URL}/verify-email`);
      }
    } else if (auth0Error.message === 'Invalid state') {
      setUser(null);
    } else {
      if (auth0Error.message) {
        alert(auth0Error.message);
      }
      setUser(null);
      logoutUser(`${process.env.NEXTAUTH_URL}/`);
    }
  }
```

After verifying the account, there is no issue faced.

**Review by mohitb35 — DISMISSED** [review]
This seems to break the signup process. After signing up as a new user with email and password, I got stuck on 

[/login?error=unauthorized&error_description=401&state=QXJfeVhkaC15R3g0SHJRVHVrckp5dkRNM0VrTlZzMmJYMmhDOWFLLTBGNQ%3D%3D](url)

instead of being directed to [/verify-email](url).

Screenshot 2022-10-05 at 12 54 55 PM

I suspect that this code in `src/features/common/Layout/Navbar/index.tsx` is not running as expected.  

```
if (auth0Error) {
    if (auth0Error.message === '401') {
      if (typeof window !== 'undefined') {
        setUser(null);
        logoutUser(`${process.env.NEXTAUTH_URL}/verify-email`);
      }
    } else if (auth0Error.message === 'Invalid state') {
      setUser(null);
    } else {
      if (auth0Error.message) {
        alert(auth0Error.message);
      }
      setUser(null);
      logoutUser(`${process.env.NEXTAUTH_URL}/`);
    }
  }
```

**Comment by norbertschuler** [comment]
> This seems to break the signup process. After signing up as a new user with email and password, I got stuck on

I needed to add support for the /verify-email for the callback in the mobiles apps. Done that now.

File: src/features/common/Layout/UserPropsContext.tsx (hunk 2 above)
Comment (norbertschuler) [reply src/features/common/Layout/UserPropsContext.tsx:19]: Not sure how to solve this.

**Pull request closed without merging**